### ISSUE TITLE
Store Gateway: Index header disk space management

### ISSUE THREAD

**Issue by yeya24** [thanos-io/thanos#7029]
### Is your proposal related to a problem?

https://github.com/thanos-io/thanos/pull/6984 adds support for lazy downloaded index headers. It can be useful if only some blocks are accessed more often than other blocks. For example, data from recent one month are usually accessed more often than data older than one year. With lazy downloaded index header feature, we can speed up store gateway start up time and index header disk space usage because we download less index headers.

However, there is no way to clean up index headers on local disk if they are not accessed anymore. If an index header file is downloaded and accessed only once, it remain on the disk forever. We only do mmap unload but there is no local data retention.

### Describe the solution you'd like

1. Purge local index headers periodically if files are not used after some time.
2. Implement size based strategy as well for ^

### Describe alternatives you've considered
NA

**Comment by pawarpranav83** [comment]
Approach - create a deleteBinary function which takes expiration time and filename, retrieves the file mod time, compares, and removes if satisfied. This function is then called by a function which runs periodically and scans all the index-header files.

Should we delete the files based on last modified time? (last access time can be obtained but it is os specific).
And should I open a PR with above changes, or we solve it using some other method? @yeya24

**Comment by douglascamata** [comment]
@yeya24 I think we can plug into the idle timeout for unloading lazy index readers. After the mmap gets removed, we would also remove the downloaded file. 

Other options: 

- Add yet one more CLI flag to specify some sort of `delete-idle-lazy-index-reader-timeout`.
- Use an LRU list and remove the least used ones. This sounds better than a size strategy to me because if an index is constantly used, the size doesn't matter -- you better keep it in the disk cache to avoid having to redownload it often.

**Comment by yeya24** [comment]
Yeah I think #7118 is a good start. But that PR would delete index headers as long as they are not accessed for idle timeout. I hope we can also check if the index header is eager downloadable or lazy downloadable and we don't clean up those blocks that are eagerly downloadable to ensure latency.

> Use an LRU list and remove the least used ones. This sounds better than a size strategy to me because if an index is constantly used, the size doesn't matter -- you better keep it in the disk cache to avoid having to redownload it often.

Yeah I think it would be a combination of LRU and size based probably. It can be a complex strategy and we can start small.

**Comment by Vidyasagar9390** [comment]
Hi @yeya24 my Thanos store disk is increasing gradually,
is there any flag to delete the unused blocks.
Do you have anything to cleanup the blocks in thanos-store.

however it keeps a small amount of information about all remote blocks on local disk and keeps it in sync with the bucket. 

but the data is increasing in the storage some times it went to 99%
